### ISSUE TITLE
VSCode invokes `provideKernels` every time we change tabs between notebooks

### ISSUE THREAD

**Issue by DonJayamanne** [microsoft/vscode#116504]
- Version: 1.54.0-insider
- OS: Darwin x64 20.2.0

Previously reported here https://github.com/microsoft/vscode/issues/114048

Steps to Reproduce:

1. Install Jupyter extension
2. Open a notebook
3. VS Code will now invoke the `provideKernels` method (this is correct)
4. Open another notebook
3. VS Code will now invoke the `provideKernels` method (this is correct)
5. Now switch between the two notebooks (go into one & then back into the other)
3. VS Code will keep invoking the `provideKernels` method

I understand #114048 was addressed.
However we are seeing some weird problems where if you were to tab between the two quickly then exeuction doesn' seem to work. Since VS Code is responsible for maintaining the kernels I believe something is going wrong there.

For instance, once a preferred kernel has been provided & the user can see a kernel on the bottom right, then why is VS Code calling the extension API yet again when we come back into this document?

This could also lead to weird bugs where the extension doesn't know about this & we send a totally different preferred kernel.
Now just by tabbing in & out of the document the preferred kernel has changed (potentially) behind the scenes & the user is unaware of this.

@rebornix @jrieken /cc

**Comment by jrieken** [comment]
> However we are seeing some weird problems where if you were to tab between the two quickly then exeuction doesn' seem to work. Since VS Code is responsible for maintaining the kernels I believe something is going wrong there.

Not quickly but as along as no cell has been executed this happens. Once you execute a cell, that kernel is "locked" to the editor instance.

**Comment by rebornix** [comment]
What @jrieken described was the fix made for #114048, only after execution happens we lock the kernel (no fetching anymore). The steps @DonJayamanne mentioned above is interesting and we should discuss about the expectation before doing any change

> 1. Install Jupyter extension
> 2. Open a notebook
> 3. VS Code will now invoke the `provideKernels` method (this is correct)
> 4. Open another notebook
> 5. VS Code will now invoke the `provideKernels` method (this is correct)
> 6. Now switch between the two notebooks (go into one & then back into the other)
> 7. VS Code will keep invoking the `provideKernels` method

Step 6 is expected, if neither of the document is dirty, meaning that when you are switching between them, the editor tab is being used, since the editor widget is reused. If the document is already dirty, we will open a new editor for the second notebook, and then you should not see that. @DonJayamanne can you confirm?

**Comment by DonJayamanne** [comment]
> If the document is already dirty, we will open a new editor for the second notebook, and then you should not see that. @DonJayamanne can you confirm?

Confirmed

**Comment by DonJayamanne** [comment]
I'm trying to optimize the following scenario:
* User opens a notebook & VSCOde invokes `providKernels`.
* Jupyter extension is slow to return all kernels
	* Assume it takes 10 seconds to return all kernels
	* & 1s to return the preferred kernel
* 90% of the time, user will have to wait for 10 seconds to run a notebook just because VS Code is fetching all of the kernels, when all it needs is just the preferred kernel.

@rebornix should I file a separate issue for this.

**Comment by jrieken** [comment]
related to https://github.com/microsoft/vscode/issues/117077. We will tackle this in March
